### DISCUSSION TITLE
Is it possible to have custom confidence threshold for individual classes during detection?

### DISCUSSION THREAD

**Discussion by ghost** [ultralytics/ultralytics#5983] [General]
So If I want customize confidence thresholds for individual classes during detection is that possible in Yolo V8? 
![image](https://github.com/ultralytics/ultralytics/assets/109454863/53b52157-1648-4e27-bfa9-ce5a17a3a91a)

**Comment by pderrenger** [comment]
@Sathyxn yes, it is possible to apply custom confidence thresholds for individual classes during detection with YOLOv8. You can achieve this by post-processing the model's predictions. After running the model's `predict` method, you can filter the results based on the class-specific confidence thresholds you have defined.

Here's a simplified example of how you might apply different confidence thresholds for each class after prediction:

```python
from ultralytics import YOLO

# Load your trained model
model = YOLO('path/to/your/model.pt')

# Perform prediction
results = model.predict(source='path/to/your/image.jpg')

# Define your custom confidence thresholds for each class
class_conf_thresholds = {0: 0.5, 1: 0.3, 2: 0.7}  # Example thresholds for three classes

# Filter results based on custom thresholds
filtered_boxes = []
for i, (box, conf, cls) in enumerate(zip(results.boxes.xyxy[0], results.boxes.conf[0], results.boxes.cls[0])):
    if conf > class_conf_thresholds[cls.item()]:
        filtered_boxes.append(box)

# Now, 'filtered_boxes' contains only the boxes that meet the custom thresholds
```

Please note that the above code is a basic example and assumes you have a model trained on a dataset with at least three classes. You would need to adjust the `class_conf_thresholds` dictionary to match the number of classes and desired thresholds for your specific use case.

**Reply by emcrobert** [reply]
Would it not be ideal to do the filtering before non-max-suppression? As you might lose some predictions during non-max-suppression(NMS) that you wouldn't have if the threshold filtering was calculated at the same time as nms.

For example say it predicts object a with confidence 0.7 and that overlaps with object b which has been predicted with confidence 0.6. If objects a's class threshold is 0.8 it's not relevant to us, but it might cause object b to be removed during NMS.

**Reply by glenn-jocher** [reply]
@emcrobert filtering before non-max suppression (NMS) can indeed help preserve relevant predictions by applying class-specific thresholds first. This approach ensures that only the desired detections influence the NMS process.
